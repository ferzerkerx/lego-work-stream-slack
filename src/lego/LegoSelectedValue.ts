import { LegoSelectedValueEntry } from './LegoSelectedValueEntry';

export class LegoSelectedValue {
  id: string;
  entries: LegoSelectedValueEntry[] = [];
}
