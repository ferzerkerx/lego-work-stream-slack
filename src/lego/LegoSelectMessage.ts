import {LegoSelectedValue} from "./LegoSelectedValue";

export class LegoSelectMessage {
  id: string;
  selectedValues: LegoSelectedValue[] = [];
  channel: string;
  date: Date = new Date();
}