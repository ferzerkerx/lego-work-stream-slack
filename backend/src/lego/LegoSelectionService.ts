import { LegoSelectMessage } from './LegoSelectMessage';
import { LegoSelectedValue } from './LegoSelectedValue';
import { LegoSelectedValueEntry } from './LegoSelectedValueEntry';

export class MessageContext {
  fullMessageId: string;
  userData: { userId: string; userName: string };
  action: any;
  channelData: { channelId: string; name: string };
}

export class LegoSelectionService {
  static createLegoSelectMessage(params: {
    legoMessage?: LegoSelectMessage;
    messageContext: MessageContext;
  }): LegoSelectMessage {
    const storedLegoMessage: LegoSelectMessage =
      params.legoMessage || new LegoSelectMessage();

    const messageContext = params.messageContext;
    let selectedValues: LegoSelectedValue[] = this._legoSelectedValues(
      storedLegoMessage.selectedValues || [],
      messageContext
    );

    return {
      id: messageContext.fullMessageId,
      selectedValues: selectedValues,
      channelData: messageContext.channelData,
      date: storedLegoMessage.date,
    };
  }

  private static _legoSelectedValues(
    currentSelectedValues: LegoSelectedValue[] = [],
    messageContext: MessageContext
  ): LegoSelectedValue[] {
    let selectedValues: LegoSelectedValue[] = JSON.parse(
      JSON.stringify(currentSelectedValues)
    );
    let currentAction = messageContext.action;

    let currentSelectedValue: LegoSelectedValue = selectedValues
      .filter(value => value.id == currentAction.name)
      .shift();

    if (!currentSelectedValue) {
      currentSelectedValue = new LegoSelectedValue();
      selectedValues.push(currentSelectedValue);
    }

    currentSelectedValue.id = currentAction.name;

    const userData = messageContext.userData;
    let legoSelectedValueEntry: LegoSelectedValueEntry = currentSelectedValue.entries
      .filter(entry => entry.userData.userId == userData.userId)
      .shift();

    if (!legoSelectedValueEntry) {
      legoSelectedValueEntry = new LegoSelectedValueEntry();
      currentSelectedValue.entries.push(legoSelectedValueEntry);
    }

    legoSelectedValueEntry.userData = userData;
    legoSelectedValueEntry.value = currentAction.selected_options[0].value;

    return selectedValues;
  }
}
