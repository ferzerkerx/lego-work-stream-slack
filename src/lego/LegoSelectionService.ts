import { LegoSelectMessage } from './LegoSelectMessage';
import { LegoSelectedValue } from './LegoSelectedValue';
import { LegoSelectedValueEntry } from './LegoSelectedValueEntry';

export class LegoSelectionService {
  static createLegoSelectMessage(params: {
    legoMessage?: LegoSelectMessage;
    fullMessageId: string;
    user: string;
    action: any;
    channel: string;
  }): LegoSelectMessage {
    const storedLegoMessage: LegoSelectMessage =
      params.legoMessage || new LegoSelectMessage();
    let selectedValues: LegoSelectedValue[] = this._legoSelectedValues(
      storedLegoMessage.selectedValues || [],
      params.user,
      params.action
    );

    return {
      id: params.fullMessageId,
      selectedValues: selectedValues,
      channel: params.channel,
      date: storedLegoMessage.date,
    };
  }

  private static _legoSelectedValues(
    currentSelectedValues: LegoSelectedValue[] = [],
    user,
    action
  ): LegoSelectedValue[] {
    let selectedValues: LegoSelectedValue[] = JSON.parse(JSON.stringify(currentSelectedValues));
    let currentAction = action;

    let currentSelectedValue: LegoSelectedValue = selectedValues
      .filter(value => value.id == currentAction.name)
      .shift();

    if (!currentSelectedValue) {
      currentSelectedValue = new LegoSelectedValue();
      selectedValues.push(currentSelectedValue);
    }

    currentSelectedValue.id = currentAction.name;

    let legoSelectedValueEntry: LegoSelectedValueEntry = currentSelectedValue.entries
      .filter(entry => entry.user == user)
      .shift();

    if (!legoSelectedValueEntry) {
      legoSelectedValueEntry = new LegoSelectedValueEntry();
      currentSelectedValue.entries.push(legoSelectedValueEntry);
    }

    legoSelectedValueEntry.user = user;
    legoSelectedValueEntry.value = currentAction.selected_options[0].value;

    return selectedValues;
  }
}
