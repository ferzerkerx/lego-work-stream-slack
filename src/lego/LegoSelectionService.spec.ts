import { LegoSelectionService, MessageContext } from './LegoSelectionService';
import { LegoSelectMessage } from './LegoSelectMessage';
import { LegoSelectedValue } from './LegoSelectedValue';

describe('LegoSelectionService', () => {
  test('createLegoSelectMessage without entries', () => {
    const storedLegoMessage: LegoSelectMessage = new LegoSelectMessage();
    storedLegoMessage.date = new Date('2018-10-01');

    const messageContext: MessageContext = {
      fullMessageId: 'someMessageId',
      userData: { userId: 'someUserId', userName: 'someuser' },
      action: { selected_options: [{ name: 'someActionName', value: 7 }] },
      channelData: { channelId: 'someChannelId', name: 'someChannel' },
    };

    const legoSelectMessage: LegoSelectMessage = LegoSelectionService.createLegoSelectMessage(
      {
        legoMessage: storedLegoMessage,
        messageContext: messageContext,
      }
    );
    expect(legoSelectMessage).toMatchSnapshot();
  });

  test('createLegoSelectMessage with existing entries', () => {
    let selectedValues: LegoSelectedValue[] = [
      {
        id: 'someActionName',
        entries: [
          {
            userData: { userId: 'someUserId', userName: 'someuser' },
            value: 8,
          },
        ],
      },
      {
        id: 'someOtherActionName',
        entries: [
          {
            userData: { userId: 'someUserId', userName: 'someuser' },
            value: 6,
          },
        ],
      },
    ];

    const storedLegoMessage: LegoSelectMessage = {
      id: 'someMessageId',
      selectedValues: selectedValues,
      date: new Date('2018-10-01'),
      channelData: { channelId: 'someChannelId', name: 'someChannel' },
    };

    const messageContext: MessageContext = {
      fullMessageId: 'someMessageId',
      userData: { userId: 'someUserId', userName: 'someuser' },
      action: { selected_options: [{ name: 'someActionName', value: 7 }] },
      channelData: { channelId: 'someChannelId', name: 'someChannel' },
    };

    const legoSelectMessage: LegoSelectMessage = LegoSelectionService.createLegoSelectMessage(
      {
        legoMessage: storedLegoMessage,
        messageContext: messageContext,
      }
    );
    expect(legoSelectMessage).toMatchSnapshot();
  });
});
