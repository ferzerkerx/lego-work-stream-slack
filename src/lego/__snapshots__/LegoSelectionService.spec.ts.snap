// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LegoSelectionService createLegoSelectMessage with existing entries 1`] = `
Object {
  "channel": "someChannel",
  "date": 2018-10-01T00:00:00.000Z,
  "id": "someMessageId",
  "selectedValues": Array [
    Object {
      "entries": Array [
        Object {
          "user": "someuser",
          "value": 7,
        },
      ],
      "id": "someActionName",
    },
    Object {
      "entries": Array [
        Object {
          "user": "someuser",
          "value": 6,
        },
      ],
      "id": "someOtherActionName",
    },
  ],
}
`;

exports[`LegoSelectionService createLegoSelectMessage without entries 1`] = `
Object {
  "channel": "someChannel",
  "date": 2018-10-01T00:00:00.000Z,
  "id": "someMessageId",
  "selectedValues": Array [
    LegoSelectedValue {
      "entries": Array [
        LegoSelectedValueEntry {
          "user": "someuser",
          "value": 7,
        },
      ],
      "id": undefined,
    },
  ],
}
`;
